File: plotMPT.py
import plotly.graph_objects as go
from statistics import mean

height = 0
nodes = []
tolerance = []
with open("MPT.csv") as f:
    for i, line in enumerate(f):
        if int(line.split(",")[0]) == height:
            nodes.append(
                (
                    float(line.split(",")[2]),
                    float(line.split(",")[3]),
                    float(line.split(",")[4]),
                    float(line.split(",")[5]),
                    round(float(line.split(",")[6]), 2),
                )
            )
            tolerance.append(nodes[-1][4])
        # if i == 10000: break

maxTol = max(tolerance)
minTol = min(tolerance)
meanTol = round(mean(tolerance), 2)

fig = go.Figure()

# Set axes properties
fig.update_xaxes(range=[-200, 200], showgrid=False, zeroline=False, mirror=True, ticks="outside", showline=True, linecolor="black")
fig.update_yaxes(range=[-100, 100], showgrid=False, zeroline=False, mirror=True, ticks="outside", showline=True, linecolor="black")

# Add shapes
for n in nodes:
    a = n[4]
    alpha = (a - minTol) / (maxTol - minTol)
    fig.add_trace(
        go.Scatter(
            mode="lines",
            x=[n[0],n[0],n[2],n[2],n[0]],
            y=[n[1],n[3],n[3],n[1],n[1]],
            fill="toself",
            name="",
            text=str(n[4]),
            fillcolor='rgba(255,143,0,'+str(alpha)+')',
            line=dict(
                color="black", width=0.25),
        )
    )

fig.add_trace(
    go.Scatter(
        x=[None],
        y=[None],
        mode='markers',
        marker=dict(
            cmin=0,
            cmax=1,
            colorscale=[[0,'rgba(255,143,0,0)'],[1, 'rgba(255,143,0,1)']],
            showscale=True,
            colorbar=dict(
                title="Node tolerance",
                tickvals=[0, meanTol, 1],
                ticktext=[str(minTol) + " (Min)", str(meanTol) + " (Mean)", str(maxTol) + " (Max)"],
                ticks="outside",
            )
        )
    )
)

# fig.update_shapes(dict(xref='x', yref='y', line=dict(color="black")))
fig.update_layout(
    showlegend=False,
    title="Tolerance map at height=0 for 1:1 read-write workload",
    xaxis_title="Longitude",
    yaxis_title="Latitude",
    font_size=6,
    plot_bgcolor="white"
)
fig.write_image(file="tMap-ST-W10-H0.pdf")

File: plotMPT.jl
using PlotlyJS

height = 0
nodes = []
tolerance = []
open("MPT.csv") do f
    for (i, line) in enumerate(eachline(f))
        vals = map(w -> parse(Float64, w), split(line, ','))
        if Int(vals[1]) == height
            t = round(vals[end]; digits = 2)
            push!(nodes, tuple(vcat(vals[3:6], t)...))
            push!(tolerance, t)
        end
    end
end

maxTol = maximum(tolerance)
minTol = minimum(tolerance)
meanTol = round(sum(tolerance) / length(tolerance); digits = 2)

traces = []
for n in nodes
    alpha = (n[5] - minTol) / (maxTol - minTol)
    push!(traces, scatter(
        x = [n[1], n[1], n[3], n[3], n[1]],
        y = [n[2], n[4], n[4], n[2], n[2]],
        fill = "toself",
        name = "",
        text = string(n[5]),
        fillcolor = string("rgba(255,143,0,", alpha, ")"),
        line = attr(
            color = "black",
            width = 0.25
        )
    ))
end

#= push!(traces, scatter(
    x=[nothing],
    y=[nothing],
    mode="markers",
    marker=attr(
        cmin=0,
        cmax=1,
        colorscale=[]
    )
)) =#

layout = Layout(
    xaxis = attr(
        range = [-200, 200],
        showgrid = false,
        zeroline = false,
        mirror = true,
        ticks = "outside",
        showline = true,
        linecolor = "black"
    ),
    yaxis = attr(
        range = [-100, 100],
        showgrid = false,
        zeroline = false,
        mirror = true,
        ticks = "outside",
        showline = true,
        linecolor = "black"
    )
    showlegend=false,
    title="Tolerance map at height=0 for 1:1 read-write workload",
    xaxis_title="Longitude",
    yaxis_title="Latitude",
    font_size=6,
    plot_bgcolor="white"
)

plot(traces, layout)
